compiler/rustc_lint/src/types.rs

```diff
@@ -597,7 +598,109 @@ struct CTypesVisitorState<'tcx> {
 enum FfiResult<'tcx> {
     FfiSafe,
     FfiPhantom(Ty<'tcx>),
-    FfiUnsafe { ty: Ty<'tcx>, reason: DiagMessage, help: Option<DiagMessage> },
+    FfiUnsafe {
+        ty: Ty<'tcx>,
+        reason: DiagMessage,
+        help: Option<DiagMessage>,
+    },
+    FfiUnsafeWrapper {
+        ty: Ty<'tcx>,
+        reason: DiagMessage,
+        help: Option<DiagMessage>,
+        wrapped: Box<FfiResult<'tcx>>,
+    },
+}
+
+/// Determine if a type is sized or not, and wether it affects references/pointers/boxes to it
+#[derive(Clone, Copy)]
+enum TypeSizedness {
+    /// sized type (pointers are C-compatible)
+    Sized,
+    /// unsized type because it includes an opaque/foreign type (pointers are C-compatible)
+    UnsizedBecauseForeign,
```

Comment:
I think I'll go with `Sizedness::UnsizedWithExternType` if you don't see a problem with it.
Because it's not just for extern types but also structs whose last member('s last member)* is an extern type.